File: AGSAPI/data/AGSDDSSpy/src/DBHandler.cpp
/* DBHandler.cpp */

#include <stdio.h>
#include <iostream>
#include <string>
#include "DBHandler.h"

DBHandler::DBHandler(std::string dbFilePath) {
  m_db = QSqlDatabase::addDatabase("QSQLITE");
  m_db.setDatabaseName(QString::fromStdString(dbFilePath));
  if (m_db.open()) {
    std::cout << "INFO :: DBHandler :: Connected to " << m_db.databaseName().toStdString() << std::endl;

    ExecQuery("PRAGMA journal_mode=WAL"); // Enables WAL mode for concurrent access to the database
    ExecQuery("DROP TABLE IF EXISTS DescObjet");
    ExecQuery("DROP TABLE IF EXISTS DDS");
    ExecQuery("CREATE TABLE DescObjet (ddsData TEXT)");
    /* ExecQuery("CREATE TABLE IF NOT EXISTS DDS (idInfo TEXT, idObjet TEXT, topicName TEXT, ddsData TEXT)"); */
    ExecQuery("CREATE TABLE DDS (pKey TEXT PRIMARY KEY, idObjet TEXT, topicName TEXT, ddsData TEXT)");

  } else {
    std::cout << "ERROR :: DBHandler :: Connection to " << m_db.databaseName().toStdString() << " failed" << std::endl;
    qDebug() << "ERROR :: DBHandler :: " << m_db.lastError().text();
  }
}

DBHandler::~DBHandler(void) {
  std::cout << "INFO :: ~DBHandler :: Closing " << m_db.databaseName().toStdString() << std::endl;
  m_db.close();
}

void DBHandler::ExecQuery(std::string queryStr) {
  QSqlQuery query;
  if (!query.exec(QString::fromStdString(queryStr))) {
    qDebug() << "ERROR :: ExecQuery :: " << query.lastError().text();
  }
}

void DBHandler::InsertDescObjetEntry(std::string ddsData) {
  QSqlQuery query;

  query.prepare("INSERT INTO DescObjet (ddsData) VALUES (:ddsData)");
  query.bindValue(":ddsData", QString::fromStdString(ddsData));

  if (!query.exec()) {
    qDebug() << "ERROR :: InsertDescObjetEntry :: " << query.lastError().text();
  }
}

void DBHandler::InsertDDSEntry(std::string idInfo, std::string idObjet, std::string topicName, std::string ddsData) {
  QSqlQuery query;

  /* DELETE */
  query.prepare("DELETE FROM DDS WHERE idInfo = :idInfo AND idObjet = :idObjet AND topicName = :topicName");
  query.bindValue(":idInfo", QString::fromStdString(idInfo));
  query.bindValue(":idObjet", QString::fromStdString(idObjet));
  query.bindValue(":topicName", QString::fromStdString(topicName));

  if (!query.exec()) {
    qDebug() << "ERROR :: InsertDDSEntry :: " << query.lastError().text();
  }

  /* INSERT */
  query.prepare("INSERT INTO DDS (idInfo, idObjet, topicName, ddsData) VALUES (:idInfo, :idObjet, :topicName, :ddsData)");
  query.bindValue(":idInfo", QString::fromStdString(idInfo));
  query.bindValue(":idObjet", QString::fromStdString(idObjet));
  query.bindValue(":topicName", QString::fromStdString(topicName));
  query.bindValue(":ddsData", QString::fromStdString(ddsData));

  if (!query.exec()) {
    qDebug() << "ERROR :: InsertDDSEntry :: " << query.lastError().text();
  }

}

void DBHandler::ReplaceDDSEntry(std::string idInfo, std::string idObjet, std::string topicName, std::string ddsData) {
  QSqlQuery query;

  /* REPLACE */
  std::string pKeySeparator = "/////";
  query.prepare("REPLACE INTO DDS (pKey, idObjet, topicName, ddsData) VALUES (:pKey, :idObjet, :topicName, :ddsData)");
  query.bindValue(":pKey", QString::fromStdString(idInfo + pKeySeparator + idObjet + pKeySeparator + topicName));
  query.bindValue(":idObjet", QString::fromStdString(idObjet));
  query.bindValue(":topicName", QString::fromStdString(topicName));
  query.bindValue(":ddsData", QString::fromStdString(ddsData));

  if (!query.exec()) {
    qDebug() << "ERROR :: ReplaceDDSEntry :: " << query.lastError().text();
  }
}

File: AGSAPI/data/AGSAPIProvider/src/DBHandler.py
#!/usr/bin/python
# coding: utf-8

from PyQt4.QtSql import QSqlDatabase, QSqlQuery
from Logger import *
import threading

class DBHandler:
    connection_id = 0

    def __init__(self, database_name):
        # Connection to the database

        # Increase connection_id for a thread-safe database access (Dirty)
        connection_id_lock = threading.Lock()
        connection_id_lock.acquire()
        self.__class__.connection_id += 1
        if self.__class__.connection_id > 1000000:
            self.__class__.connection_id = 0
        self.instance_connection_id = str(self.__class__.connection_id)
        connection_id_lock.release()

        self.db = QSqlDatabase.addDatabase("QSQLITE", str(self.instance_connection_id))

        self.db.setDatabaseName(database_name)
        if self.db.open():
            print("Connected to the database ID->" + self.instance_connection_id)
            pass
        else:
            print("Cannot connect to the database")

    def close(self):
        print("Closing connection to the database ID->" + self.instance_connection_id)
        self.db.close()
        del self.db
        QSqlDatabase.removeDatabase(self.instance_connection_id)

    def get_descobjet_list(self):
        query = QSqlQuery(self.db)
        if not query.exec_("SELECT * FROM DescObjet"):
            logging.error("%s", query.lastError().text())

        descobjet_list = []
        while query.next():
            dds_data = unicode(query.record().value(0).toString())
            descobjet_list.append(dds_data)

        return descobjet_list

    def get_dds_data_list(self, id_objet, topic_name):
        query = QSqlQuery(self.db)
        
        # "DATA LIKE" to match things like AlarmeV2 even when the topic_name passed is Alarme
        query.prepare('SELECT * FROM DDS WHERE idObjet = :idObjet AND topicName LIKE "{0}%"'.format(topic_name))
        query.bindValue(":idObjet", id_objet)
        # query.bindValue(":topicName", topic_name)

        if not query.exec_():
            logging.error("%s", query.lastError().text())

        dds_data_list = []
        while query.next():
            dds_data = unicode(query.record().value(3).toString())
            dds_data_list.append(dds_data)

        return dds_data_list
